### ISSUE TITLE
Add better support for re-using named bind values

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3973]
One of the advantages of named bind values is the fact that they can be re-used in statements:

```sql
INSERT INTO t (id, val1, val2)
VALUES (1, :val, :val)
```

While the jOOQ API generally allows for re-using the same bind variable name in a single statement, there is currently not enough means (short of implementing a `BindContext`) of extracting the individual locations of the named `Param`s by index, prior to binding the actual values to a `PreparedStatement`.

In other words, the repetitive variable name cannot be translated to indexes `1` and `2` when binding to JDBC.
